File: EPI/19.2/typescript/19.2.ts
export interface Cell {
    row: number;
    col: number;
}

export function flip(grid: boolean[][], cell: Cell): void {
    const color = grid[cell.row][cell.col];
    grid[cell.row][cell.col] = !color;

    for (const neighbor of neighbors(grid, cell)) {
        if (grid[neighbor.row][neighbor.col] === color) {
            flip(grid, neighbor);
        }
    }
}

function neighbors(grid: boolean[][], {row, col}: Cell): Cell[] {
    return [
        {row: row - 1, col},
        {row, col: col + 1},
        {row: row + 1, col},
        {row, col: col - 1},
    ].filter(({row: r, col: c}) =>
        r >= 0 && r < grid.length &&
        c >= 0 && c < grid[r].length
    );
}

File: EPI/Graphs/Paint a Boolean matrix/swift/19.2.swift
public enum Color: Int {
    case White = 0
    case Black = 1
}

public typealias Coord = (row: Int, col: Int)

public func flipRegion(_ cell: Coord, in image: inout [[Color]]) {
    let color = image[cell.row][cell.col]
    image[cell.row][cell.col] = color == .White ? .Black : .White

    for neighbor in neighbors(cell, in: image) where image[neighbor.row][neighbor.col] == color {
        flipRegion(neighbor, in: &image)
    }
}

private func neighbors(_ cell: Coord, in image: [[Color]]) -> [Coord] {
    let (row, col) = cell

    let potentialNeighbors = [
        (row - 1, col),
        (row + 1, col),
        (row, col - 1),
        (row, col + 1),
    ]

    return potentialNeighbors.filter { (r, c) in
        r >= 0 && r < image.count &&
        c >= 0 && c < image[r].count
    }
}
